### PR TITLE
Update ticket_holder table schema

### LINKED ISSUE #6669: Make ticket_id and event_id not null in TicketHolder
Self explanatory

### PR DESCRIPTION
**Pull request by ashwani99**
Fixes #6669 

#### Short description of what this resolves:
Updated the fields ticket_id and event_id to be not null

#### Changes proposed in this pull request:

- Updated schema of `ticket_holder` table. The fields `ticket_id` and `event_id` have been added with not null contraint

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

File: app/factories/attendee.py [hunk 1]
```diff
@@ -11,6 +12,7 @@ class Meta:
         sqlalchemy_session = db.session
 
     event = factory.RelatedFactory(EventFactoryBasic)
+    ticket = factory.RelatedFactory(TicketFactory)
```
Comment (codedsun) [inline app/factories/attendee.py]: why?

File: app/factories/attendee.py [hunk 2]
```diff
@@ -21,4 +23,4 @@ class Meta:
     is_checked_in = True
     pdf_url = common.url_
     event_id = 1
-    ticket_id = None
```
Comment (codedsun) [inline app/factories/attendee.py]: Not needed

**Comment by vishakhanihore** [comment]
Migration is needed for this.

File: app/factories/attendee.py (hunk 1 above)
Comment (ashwani99) [reply app/factories/attendee.py]: Without doing this, while running tests, some test were failing because of `ticket_id` being `None`

Comment (codedsun) [reply app/factories/attendee.py]: Because you made ticket_id nullabse = false, that's why

File: app/factories/attendee.py [hunk 3]
```diff
@@ -21,4 +21,5 @@ class Meta:
     is_checked_in = True
     pdf_url = common.url_
     event_id = 1
+    # ticket_id = 1
```
Comment (kushthedude) [inline app/factories/attendee.py]: Remove comment

File: app/models/ticket_holder.py [hunk 4]
```diff
@@ -37,15 +37,15 @@ class TicketHolder(SoftDeletionModel):
...
+    ticket_id: int = db.Column(db.Integer, db.ForeignKey('tickets.id', ondelete='CASCADE'), nullable=False)
     order_id: int = db.Column(db.Integer, db.ForeignKey('orders.id', ondelete='CASCADE'))
     is_checked_in: bool = db.Column(db.Boolean, default=False)
     is_checked_out: bool = db.Column(db.Boolean, default=False)
     device_name_checkin: str = db.Column(db.String)
     checkin_times: str = db.Column(db.String)
     checkout_times: str = db.Column(db.String)
     attendee_notes: str = db.Column(db.String)
-    event_id: int = db.Column(db.Integer, db.ForeignKey('events.id', ondelete='CASCADE'))
+    event_id: int = db.Column(db.Integer, db.ForeignKey('events.id', ondelete='CASCADE'), nullable=False)
```
Comment (kushthedude) [inline app/models/ticket_holder.py]: Add migration files for the following.

File: docs/api/blueprint/attendees.apib [hunk 5]
```diff
@@ -73,7 +73,8 @@ Create a new attendee.
                     "country": "IN",
                     "is-checked-in": "false",
                     "attendee-notes": "example",
-                    "pdf-url": "http://example.com"
+                    "pdf-url": "http://example.com",
```
Comment (kushthedude) [inline docs/api/blueprint/attendees.apib]: Revert

File: docs/api/blueprint/attendees.apib [hunk 6]
```diff
@@ -73,7 +73,8 @@ Create a new attendee.
...
+                    "ticket-id": "1"
```
Comment (kushthedude) [inline docs/api/blueprint/attendees.apib]: revert

File: app/factories/attendee.py [hunk 7]
```diff
@@ -21,4 +23,4 @@ class Meta:
     is_checked_in = True
     pdf_url = common.url_
     event_id = 1
-    ticket_id = None
+    ticket_id = 1
```
Comment (kushthedude) [inline app/factories/attendee.py]: revert

Comment (iamareebjamal) [reply app/factories/attendee.py]: Can't be reverted. You cannot create an attendee without ticket ID now, and you shouldn't be able to

File: docs/api/blueprint/attendees.apib (hunk 6 above)
Comment (ashwani99) [reply docs/api/blueprint/attendees.apib]: @kushthedude If you explain the problem it would be very helpful

File: tests/hook_main.py [hunk 8]
```diff
@@ -1858,7 +1858,7 @@ def attendee_post(transaction):
         ticket = TicketFactory()
         db.session.add(ticket)
 
-        attendee = AttendeeFactory(ticket_id=1)
+        attendee = AttendeeFactory(ticket_id=1, event_id=1)
```
Comment (iamareebjamal) [inline tests/hook_main.py]: It's the default value. Not needed

File: docs/api/blueprint/attendees.apib [hunk 9]
```diff
@@ -73,7 +73,9 @@ Create a new attendee.
                     "country": "IN",
                     "is-checked-in": "false",
                     "attendee-notes": "example",
-                    "pdf-url": "http://example.com"
+                    "pdf-url": "http://example.com",
+                    "ticket-id": "1",
+                    "event-id": "1"
```
Comment (iamareebjamal) [inline docs/api/blueprint/attendees.apib]: Unneeded

File: app/api/schema/attendees.py [hunk 10]
```diff
@@ -50,7 +50,6 @@ def validate_json(self, data, original_data):
     gender = fields.Str(allow_none=True)
     birth_date = fields.DateTime(allow_none=True)
 
-    ticket_id = fields.Str(allow_none=True)
```
Comment (prateekj117) [inline app/api/schema/attendees.py]: Undo this and run the tests.

Comment (prateekj117) [reply app/api/schema/attendees.py]: @iamareebjamal I think the tests are failing due to this.

Comment (ashwani99) [reply app/api/schema/attendees.py]: I have tried it. The issue still persists.

File: app/api/attendees.py [hunk 11]
```diff
@@ -74,6 +74,11 @@ def before_post(self, args, kwargs, data):
                 {'pointer': '/data/relationships/ticket'},
                 "Ticket belongs to a different Event"
             )
+
+        # populate event id and ticket id
+        data['ticket_id'] = ticket.id
+        data['event_id'] = ticket.event_id
```
Comment (iamareebjamal) [inline app/api/attendees.py]: Not needed

Comment (ashwani99) [reply app/api/attendees.py]: Hi @iamareebjamal I am trying to debug what is the exact reason of the `ticket_id` and `event_id`  becoming `null` after the object creation. That's why I tried passing them in the data payload. Any you help me find out what can be the exact reason? Am I going in the right direction as per tracing the problem?

File: app/api/schema/attendees.py [hunk 12]
```diff
@@ -93,7 +93,7 @@ class Meta:
                           self_view_kwargs={'id': '<id>'},
                           related_view='v1.ticket_detail',
                           related_view_kwargs={'attendee_id': '<id>'},
-                          schema='TicketSchemaPublic',
+                          schema='TicketSchema',
```
Comment (iamareebjamal) [inline app/api/schema/attendees.py]: Can you please stop hit and trial. This is not needed and wrong

**Comment by iamareebjamal** [comment]
Stale

**Pull request closed without merging**